File: bench/kernels/d2_substr_search/ref.c
/* d2-substr-search (family D), NAIVE C: sliding `memcmp` over the same
 * haystack. Deliberately not `strstr` — see bench/protocol.md on comparing
 * hand loops to hand-vectorized library routines.
 * Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static const char CHUNK[] = "the pack moves at dusk and the wolf waits for nothing at all ";
#define REPS 9000
#define M 5

static int64_t count_occurrences(const char *hay, size_t len, const char *needle) {
    int64_t hits = 0;
    for (size_t i = 0; i + M <= len; i++)
        if (memcmp(hay + i, needle, M) == 0) hits++;
    return hits;
}

int main(int argc, char **argv) {
    int64_t ops = argc > 1 ? (int64_t)strtoull(argv[1], 0, 10) : 20;
    size_t clen = strlen(CHUNK);
    size_t len = clen * REPS;
    char *buf = malloc(len + 1);
    for (int i = 0; i < REPS; i++) memcpy(buf + (size_t)i * clen, CHUNK, clen);
    buf[len] = 0;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    int64_t sink = 0;
    for (int64_t k = 0; k < ops; k++) sink = count_occurrences(buf, len, "wolf ");
    clock_gettime(CLOCK_MONOTONIC, &t1);
    free(buf);
    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000ull
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    printf("{\"ns\":%llu,\"ops\":%llu,\"sink\":%lld}\n",
           (unsigned long long)ns, (unsigned long long)((uint64_t)ops * len),
           (long long)sink);
    return 0;
}

File: bench/kernels/d2_substr_search/ref.rs
//! d2-substr-search (family D), Rust reference: sliding slice compare over
//! the same haystack (not `str::find`, per bench/protocol.md).
//! Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}.

use std::hint::black_box;
use std::time::Instant;

const CHUNK: &str = "the pack moves at dusk and the wolf waits for nothing at all ";
const REPS: usize = 9000;
const M: usize = 5;

fn count_occurrences(hay: &[u8], needle: &[u8]) -> i64 {
    let mut hits: i64 = 0;
    let mut i = 0usize;
    while i + M <= hay.len() {
        if &hay[i..i + M] == needle {
            hits += 1;
        }
        i += 1;
    }
    hits
}

fn main() {
    let ops: u64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(20);
    let hay = CHUNK.repeat(REPS);
    let bytes = hay.as_bytes();
    let t = Instant::now();
    let mut sink: i64 = 0;
    for _ in 0..ops {
        sink = count_occurrences(black_box(bytes), b"wolf ");
    }
    let ns = t.elapsed().as_nanos();
    println!(
        "{{\"ns\":{ns},\"ops\":{},\"sink\":{sink}}}",
        ops * bytes.len() as u64
    );
}
